### PR TITLE
ENH: Improve TopHatImageFilter classes coverage

### PR DESCRIPTION
**Pull request by jhlegarreta**
Improve `itk::BlackTopHatImageFilter` and `itk::WhiteTopHatImageFilter`
classes coverage:

- Exercise the basic object methods using the
  `ITK_EXERCISE_BASIC_OBJECT_METHODS` macro.
- Use the `ITK_TRY_EXPECT_NO_EXCEPTION` macro when updating the filters.
- Exercise the Set/Get methods using the `ITK_TEST_SET_GET_VALUE` macro.
- Refactor the test to reduce boilerplate code.
- Add arguments to the test to improve the class testing of the additional
  ivars.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

**Comment by jhlegarreta** [comment]
Looks like a number of tests were not run in the `Circle CI` and `ITK.Linux` `Azure Pipelines` tests for reasons that I'd dare to say are not linked to the changes in this PR.

This PR was based on the latest `master` commit just before pushing, so re-triggering the CIs may do the trick.

**Comment by dzenanz** [comment]
There are compile errors, they seem to be the same in [CircleCI](https://open.cdash.org/viewBuildError.php?buildid=6743923) and [Azure](https://open.cdash.org/viewBuildError.php?buildid=6744096).

File: Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx [hunk 1]
```diff
@@ -24,14 +24,68 @@
...
+                               typename TKernelImageFilter::KernelType ball,
+                               const bool                              safeBorder,
+                               const int                               algorithm,
+                               const bool                              forceAlgorithm,
+                               const char *                            inputFileName,
+                               const char *                            outputFileName)
+{
+  // Declare the reader and writer
+  using ReaderType = itk::ImageFileReader<TKernelImageFilter::InputImageType>;
+  ReaderType::Pointer reader = ReaderType::New();
```
Comment (dzenanz) [inline Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx:40]: `ReaderType::Pointer` needs to be `typename ReaderType::Pointer`

File: Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx [hunk 2]
```diff
@@ -24,14 +24,68 @@
...
+
+  // Connect the pipeline
+  filter->SetInput(reader->GetOutput());
+
+  ITK_TRY_EXPECT_NO_EXCEPTION(filter->Update());
+
+
+  // Write output
+  using WriterType = itk::ImageFileWriter<TKernelImageFilter::OutputImageType>;
+  WriterType::Pointer writer = WriterType::New();
```
Comment (dzenanz) [inline Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx:63]: Also add `typename` here.

**Review by dzenanz — COMMENTED** [review]
If a type is defined within a class depending on a template, then it needs to be prefixed by `typename` keyword. Visual Studio does not enforce this, but GCC does.

File: Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx (hunk 1 above)
Comment (jhlegarreta) [reply Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx]: Fixed in 18d1de6. Thanks for spotting these !

File: Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx (hunk 2 above)
Comment (jhlegarreta) [reply Modules/Filtering/MathematicalMorphology/test/itkTopHatImageFilterTest.cxx]: Fixed in 18d1de6. Thanks for spotting these !

**Comment by jhlegarreta** [comment]
> There are compile errors, they seem to be the same in CircleCI and Azure.

@dzenanz Sorry, I missed those: did not realize that tests not being run was derived from compilation errors 🤦, and looks like I did not read closely enough the build raw logs. Warnings/errors in `CDash` are easier to spot :-) ! Thanks :+1:.

**Comment by dzenanz** [comment]
CDash link has been broken for a while now. Workaround is to go to dfgddgf and search for PR number (in this case `PR1986`). It looks like you [missed](https://open.cdash.org/viewBuildError.php?buildid=6744096) some instances of `typename`.

P.S. you just made a force-push which might have addressed them 😄

**Comment by dzenanz** [comment]
Also thanks for improving tests!

**Comment by dzenanz** [comment]
Message from admin:
> We are going to reboot mysql1 at 4:30pm today. It should only take 5-10 minutes. It will affect all cdash instances while the database is down.

This is US Eastern time zone.

**Comment by jhlegarreta** [comment]
The reported failure of `ITK.macOS` is unrelated: `itkLSMImageIOTest`:
```

2020-09-04T22:10:16.6843470Z           Start  959: itkLSMImageIOTest
2020-09-04T22:10:16.6845240Z 
2020-09-04T22:10:16.6847790Z 959: Test command:
/Users/runner/work/1/s-build/bin/ITKIOLSMTestDriver
"--ignoreInputInformation"
"--compareCoordinateTolerance" "1.0e-6" "--compareDirectionTolerance"
"1.0e-6"
"--compare"
"/Users/runner/work/1/s-build/ExternalData/Testing/Data/Baseline/IO/cthead1.tif"
"/Users/runner/work/1/s-build/Testing/Temporary/cthead1.tif"
"itkLSMImageIOTest"
"/Users/runner/work/1/s-build/ExternalData/Testing/Data/Input/cthead1.lsm"
"/Users/runner/work/1/s-build/Testing/Temporary/cthead1.tif"
(...)
Exception detected while reading /Users/runner/work/1/s-build/ExternalData/Testing/Data/Baseline/IO/cthead1.tif:
Could not create IO object for reading file /Users/runner/work/1/s-build/Testing/Temporary/cthead1.tif
```

**Review by dzenanz — APPROVED** [review]

**Review by dzenanz — APPROVED** [review]

**Pull request merged**